KAS-3579: Minimal `include`s

sergiofenoll opened the pull request:
Not the most efficient function, but it should suffice. To apply the minimal includes to all outgoing queries, the `application` adapter has two overriden methods, `buildQuery` and `buildURL`. It's not entirely clear to me based on the docs which method is applied to which store calls, but while testing it seems that `buildURL` is only used for certain methods, e.g. `findRecord`, while `buildQuery` is only used for others, e.g. `query`. So here both methods are overriden just to be sure.

sergiofenoll committed:
feat: add minimalInclude function

This function returns the minimal number of paths of a given list of
paths for the include parameter of a request to mu-cl-resources.

Paths that are full prefixes of another path get deleted, as they are
unnecessary.

sergiofenoll committed:
feat: apply minimalInclude to all outgoing requests

sergiofenoll committed:
docs: document a query that is cached by the cache-warmup service

sergiofenoll committed:
docs: document the minimalInclude function a little bit

sergiofenoll committed:
fix: use startsWith() instead of includes() to check for prefixes

erikap commented on app/utils/minimal-include.js:
```diff
@@ -0,0 +1,38 @@
+          isPrefix = otherPath.startsWith(`${path}.`)
+        }
+      });
+      if (isPrefix) return acc;
+      return [...acc, path];
```
`return isPrefix ? acc : [...acc, path];` to improve readability?

erikap commented on app/utils/minimal-include.js:
```diff
@@ -0,0 +1,38 @@
+ * doesn't need to be part of the include parameter, so it gets removed by this
+ * function.
+ *
+ * The function works by iterating over each path and storing only paths that
+ * aren't a prefix of other paths. For each path, we iterate over all other
+ * paths and make sure that (1) the current path is not a prefix of another path
+ * and (2) that other paths are not full prefix of the current path. If (1), we
+ * discard the current path, if (2) we discord the other path.
```
Can we maybe simplify by first sorting the `paths` array on string length? As such we only need to check (1) as (2) cannot occur then.

erikap commented on app/utils/minimal-include.js:
```diff
@@ -0,0 +1,38 @@
+      if (isPrefix) return acc;
+      return [...acc, path];
+    }, [])
+    .join(',');
```
Can we do an alphabetical sort of the resulting array in the end to 'stabilize' the outcome of the function?

sergiofenoll replied:
I initially didn't want to depend on sorting, just in case the cache-warmup had non-sorted paths so that I didn't have to change them. Since it doesn't, it shouldn't matter whether we sort or not. Perhaps it's even better to just sort in the beginning on the paths (e.g. `agendaitem, agendaitem-type.kind, agendaitem.status` → `agendaitem, agendaitem.status, agendaitem-type.kind`), then the next step is just a single pass over the sorted list to check if a path is a prefix of the next path?

sergiofenoll committed:
refactor: re-work minimalInclude function using sorting

First sort, then eliminate prefixes. Should make the function simpler,
and ensures a stable output.

erikap committed:
Merge branch 'development' into feature/KAS-3579-minimal-includes
